runners/spark/src/main/java/org/apache/beam/runners/spark/io/SparkUnboundedSource.java

```diff
@@ -105,6 +112,23 @@ public Metadata call(Tuple2<Iterable<byte[]>, Metadata> t2) throws Exception {
           }
         });
 
+    // Accumulate unbounded source metrics
+    metadataDStream.foreachRDD(new VoidFunction<JavaRDD<Metadata>>() {
+      @Override
+      public void call(JavaRDD<Metadata> metadataRDD) throws Exception {
+        final Accumulator<SparkMetricsContainer> metricsAccum = MetricsAccumulator.getInstance();
+        metadataRDD.foreachPartition(new VoidFunction<Iterator<Metadata>>() {
+          @Override
+          public void call(Iterator<Metadata> metadatas) throws Exception {
+            while (metadatas.hasNext()) {
```

Comment:
Since the "Metadata stream" is being materialized anyway by `ReportingDStream`, how about you simply do this there ? instead of adding another branch to the graph ?